// datocms-client/lib/wpImport/import/categories.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = categories;

var _allPages = _interopRequireDefault(require("../utils/allPages"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _require = require('../utils/progress'),
    progress = _require.progress;

function categories(_x, _x2, _x3) {
  return _categories.apply(this, arguments);
}

function _categories() {
  _categories = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(dato, wp, itemTypeId) {
    var mapping, resources, tick, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, category, childCategories, tick2, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, _category;

    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            mapping = {};
            _context.next = 3;
            return (0, _allPages.default)('Fetching categories', wp.categories());

          case 3:
            resources = _context.sent;
            tick = progress('Creating categories', resources.length);
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            _context.prev = 8;
            _iterator = resources[Symbol.iterator]();

          case 10:
            if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
              _context.next = 18;
              break;
            }

            category = _step.value;
            _context.next = 14;
            return tick(category.name, dato.items.create({
              itemType: itemTypeId,
              name: category.name,
              slug: category.slug,
              description: category.description
            }));

          case 14:
            mapping[category.id] = _context.sent.id;

          case 15:
            _iteratorNormalCompletion = true;
            _context.next = 10;
            break;

          case 18:
            _context.next = 24;
            break;

          case 20:
            _context.prev = 20;
            _context.t0 = _context["catch"](8);
            _didIteratorError = true;
            _iteratorError = _context.t0;

          case 24:
            _context.prev = 24;
            _context.prev = 25;

            if (!_iteratorNormalCompletion && _iterator.return != null) {
              _iterator.return();
            }

          case 27:
            _context.prev = 27;

            if (!_didIteratorError) {
              _context.next = 30;
              break;
            }

            throw _iteratorError;

          case 30:
            return _context.finish(27);

          case 31:
            return _context.finish(24);

          case 32:
            childCategories = resources.filter(function (c) {
              return c.parent;
            });
            tick2 = progress('Reordering categories', childCategories.length);
            _iteratorNormalCompletion2 = true;
            _didIteratorError2 = false;
            _iteratorError2 = undefined;
            _context.prev = 37;
            _iterator2 = childCategories[Symbol.iterator]();

          case 39:
            if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
              _context.next = 46;
              break;
            }

            _category = _step2.value;
            _context.next = 43;
            return tick2(_category.name, dato.items.update(mapping[_category.id], {
              parentId: mapping[_category.parent],
              position: 0
            }));

          case 43:
            _iteratorNormalCompletion2 = true;
            _context.next = 39;
            break;

          case 46:
            _context.next = 52;
            break;

          case 48:
            _context.prev = 48;
            _context.t1 = _context["catch"](37);
            _didIteratorError2 = true;
            _iteratorError2 = _context.t1;

          case 52:
            _context.prev = 52;
            _context.prev = 53;

            if (!_iteratorNormalCompletion2 && _iterator2.return != null) {
              _iterator2.return();
            }

          case 55:
            _context.prev = 55;

            if (!_didIteratorError2) {
              _context.next = 58;
              break;
            }

            throw _iteratorError2;

          case 58:
            return _context.finish(55);

          case 59:
            return _context.finish(52);

          case 60:
            return _context.abrupt("return", mapping);

          case 61:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this, [[8, 20, 24, 32], [25,, 27, 31], [37, 48, 52, 60], [53,, 55, 59]]);
  }));
  return _categories.apply(this, arguments);
}
